crates/core/tedge_api/src/entity_store.rs

```diff
@@ -0,0 +1,478 @@
+//! A store containing registered MQTT entities.
+//!
+//! References:
+//!
+//! - <https://github.com/thin-edge/thin-edge.io/issues/2081>
+//! - thin-edge documentation: Reference > MQTT API > Entity Registration >
+//!   Entity store
+
+use std::collections::HashMap;
+
+use mqtt_channel::Message;
+use mqtt_channel::Topic;
+
+type EntityId = String;
+// type checker is not able to infer `&EntityId` as &str, so alias is needed
+// TODO: use copy type for IDs
+type EntityIdRef<'a> = &'a str;
+
+/// A store for topic-based entity metadata lookup.
+///
+/// This object is a hashmap from MQTT topics to entities (devices or services)
+/// that publish on those topics. It keeps track of type of entities, their
+/// relationships (parent and child entities), and other metadata.
+///
+/// The entity store takes as input registration messages published by entities
+/// (devices and services) and stores information about entities and their
+/// hierarchy, allowing to efficiently query it. It's possible to:
+///
+/// - enumerate all registered devices
+/// - check if a given entity is already registered
+/// - query services and child devices of a given device
+/// - query parent of an entity
+///
+/// # Examples
+///
+/// ```
+/// # use tedge_mqtt_ext::{Message, Topic};
+/// # use c8y_mapper_ext::entity_store::{EntityStore, EntityRegistrationMessage};
+/// let mqtt_message = Message::new(
+///     &Topic::new("te/device/main//").unwrap(),
+///     r#"{"@type": "device"}"#.to_string(),
+/// );
+/// let registration_message = EntityRegistrationMessage::try_from(mqtt_message).unwrap();
+///
+/// let mut entity_store = EntityStore::new();
+///
+/// entity_store.update(registration_message).unwrap();
+/// assert_eq!(entity_store.main_device(), Some("te/device/main//"));
+/// ```
+#[derive(Debug, Clone)]
+pub enum EntityStore {
+    Uninitialised,
+    Initialised {
+        main_device: EntityId,
+        entities: HashMap<EntityId, EntityMetadata>,
+    },
+}
```

Comment:
I talked with @reubenmiller about whether the main device also has to be registered, and whether or not it can be auto-registered, and I ended up rewriting this once again under the assumption that when the entity store is created, the main device is already known.